SplunkRumWorkspace/SplunkRum/SplunkRum/SessionBasedSampling.swift

```diff
@@ -0,0 +1,78 @@
+//
+/*
+Copyright 2021 Splunk Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import Foundation
+import OpenTelemetrySdk
+import OpenTelemetryApi
+
+class SessionBasedSampling {
+    static var probability: Double = 1.0
+    static var sessionCount: Int = 0
+    static var timer: Timer?
+    init(ratio: Double) {
+        SessionBasedSampling.probability = ratio
+    }
+    /**check that session will be sampled or not**/
+    public class func sessionshouldSample() {
+        let sampling_percentage = SessionBasedSampling.probability
+        let step = Double(1/sampling_percentage)
+
+        let roundedStepValue = round(step * 10) / 10.0
+        var result = false
+        switch SessionBasedSampling.probability {
+        case 0.0:
+            result = false
+        case 1.0:
+            result = true
+
+        default:
+            if floor(Double(SessionBasedSampling.sessionCount).truncatingRemainder(dividingBy: roundedStepValue)) == 0 {  // if SessionBasedSampling.sessionCount % Int(step) == 0 {
+                result = true
+            } else {
+                result = false
+            }
+        }
+
+        var psampler: Sampler
+        if result {
+                psampler = OpenTelemetrySdk.Samplers.parentBased(root: Samplers.alwaysOn, remoteParentSampled: Samplers.alwaysOn, remoteParentNotSampled: Samplers.alwaysOn, localParentSampled: Samplers.alwaysOn, localParentNotSampled: Samplers.alwaysOn)
+        } else {
+              psampler = OpenTelemetrySdk.Samplers.parentBased(root: Samplers.alwaysOff, remoteParentSampled: Samplers.alwaysOff, remoteParentNotSampled: Samplers.alwaysOff, localParentSampled: Samplers.alwaysOff, localParentNotSampled: Samplers.alwaysOff)
```

Comment:
Wouldn't it be easier to use the `Sampler` protocol and set session based sampler as the active sampler in tracer provider?

Reply:
Yes, I had tried to use `alwaysOn` and `alwaysOff` sampler. But once active sampler’s value be `alwaysOff`, no session was sampled after that even though I changed value of active sampler to `alwaysOn`. As a solution , I use `parent based `sampler protocol in place of it.